training/CHAPEL_TUTORIAL.md

```diff
@@ -0,0 +1,861 @@
+```
+<a id="ex11"></a>
+#### Example 11: Ternary and `forall` Expression
+We can combine ternary with other tools such as forall expressions.
+```chapel
+writeln([i in 0..#10] if i%2 == 0 then i+10 else -100);
+```
+```console
+10 -100 12 -100 14 -100 16 -100 18 -100
+```
+
+<a id="TIY2"></a>
+#### Try It Yourself 2: Array Absolute Value
+Problem:
+Write a `proc` using a ternary to take an `int array`, `A`, 
+and return an array where index `i` is the absolute value of `A[i]`
+
+Input: 
+`[-3, 7, 0, -4, 12]`
+
+<details>
+  <summary>Potential Solution</summary>
+
+```Chapel
+proc arrayAbsVal(A) {
+  return [a in A] if a >= 0 then a else -a;
+}
+```
+</details>
+Expected Output:
+
+```console
+3 7 0 4 12
+```
+
+<a id="introspection"></a>
+## Introspection
+Introspection is determining properties of a function argument at runtime.
+This is often used to determine the type and/or domain of a function argument.
+Introspection can avoid duplicating a `proc` for multiple types when none of the logic has changed.
+Using introspection will result in a [generic](https://chapel-lang.org/docs/language/spec/generics.html#generics) function.
+
+The syntax for this is:
+```Chapel
+proc foo(array: [?D] ?t, value: t, array2: [?D2] ?t2) {
+  // using `D` and `t` in this proc will refer to `array`'s domain and type respectively
+  // `value` has the same type as `array` so there's no need for the `?`
+  // `D2` and `t2` refer to the domain and type of `array2`
+}
+```
+<a id="ex12"></a>
+#### Example 12: Introspection
+If we adapt our absolute value example to use introspection,
+we can use the same `proc` for `real array`s and `int array`s.
+```Chapel
+proc absoluteVal(a: [?D] ?t): [D] t {
+  return [i in D] if a[i] >= 0 then a[i] else -a[i];
+}
+
+var r: [0..#5] real = [-3.14, 7:real, 0.0, INFINITY, -INFINITY];
+writeln(absoluteVal(r));
+
+var i: [0..#5] int = [-3, 7, 0, -4, 12];
+writeln(absoluteVal(i));
+```
+```console
+3.14 7.0 0.0 inf inf
+3 7 0 4 12
+```
+
+<a id="promotion"></a>
+## Promotion
+[Promotion](https://chapel-lang.org/docs/users-guide/datapar/promotion.html) is a way to obtain data parallelism implicitly.
+
+A function or operation that operates on a type can automatically work on an array of that type
+by essentially applying the function to every element of the array.
+
+<a id="ex13"></a>
+#### Example 13: Promotion
+Returning to the factorial example, our `proc` can operate on an `int array` even though it's only defined to accept an `int`.
+```chapel
+proc factorial(n: int) {
+  return * reduce [i in 1..#n] i;
+}
+writeln(factorial(5));
+writeln(factorial([1, 2, 3, 4, 5]));
+```
+```console
+120
+1 2 6 24 120
+```
+
+<a id="zip_and_agg"></a>
+## Zippered Iteration and Aggregation
+
+<a id="zippered_iteration"></a>
+### Zippered Iteration
+[Zippered Iteration](https://chapel-lang.org/docs/users-guide/base/zip.html#zippered-iteration) is simultaneously
+iterating over multiple iterables (usually arrays and ranges) with compatible domains.
```

Comment:
More generally, "with compatible shape and size" (since ranges don't have domains).  You could then add something like "for arrays, this means that their domains have the same number of elements in each dimension."